### ISSUE TITLE
Feature request to block whole subnets instead of individual IP addresses

### ISSUE THREAD

**Issue by alanmels** [backdrop-contrib/ip_blocking#20]
The Protected Forms module works very well with IP blocking module to automatically block all the offending IP addresses as shown on the screenshot below:
 
<img width="1162" alt="Screenshot 2024-04-26 at 11 53 53 PM" src="https://github.com/backdrop-contrib/ip_blocking/assets/56604405/5fd60cf9-7b78-4970-9dfa-5873b2c959a5">

However, as you see the malicious parties keep trying just by slightly changing the IP addresses, hence are the following questions:

1. We use the following function to send the offending IP addresses to block-list:

```php
/**
 * Helper function that bans the offending IP address.
 */
function _ban_ip_address() {
  global $user;
  // Prevent admins from getting blocked.
  if (user_access('administer protected forms', $user) == FALSE) {
    $ip_address = ip_address();
    // Count how many times user was caught entering spam.
    $result = db_select('watchdog', 'w')
      ->fields('w')
      ->condition('type', 'protected forms', '=')
      ->condition('hostname', $ip_address, '=')
      ->execute();
    // Block if user exceeded the set limit.
    $ban_threshold = config_get('protected_forms.settings', 'ban_threshold');
    if ($result->rowCount() >= $ban_threshold) {
      db_insert('blocked_ips')
        ->fields(array('ip' => $ip_address, 'reason' => 'Blocked by Protected Forms for spamming.', 'time' => time(), 'uid' => $user->uid))
        ->execute();
      watchdog('protected forms', "IP address <em>$ip_address</em> is blocked for reaching the ban threshold.");
    }
  }
}
```

Would the module's maintainer consider implementing a simple API command to let other modules interact with IP blocking module using just one command, something like `ip_blocking_block($ip_address)`?

2. Is it possible for the IP blocking module to start supporting blocking the whole subnets  instead of individual IP addresses. For example, all those offending IP addresses could be blocked with a single command `ip_blocking_block('172.68.190.0/24')` by indicating the whole subnet.

3. Would be very nice to have an `edit` option next to exiting `unblock | check` options on the `admin/config/people/ip-blocking`, so admins could manually change IP addresses.

Thanks for consideration!

**Comment by findlabnet** [comment]
Hi Alan, your use case is very close to one used in the 'Antiscan' module, which defines additional reasonable big uid for use in DB records. There is definitely room for some generic API calls.

2. About blocking entire subnets instead of individual IP addresses.
Some users reported having 80K+ blocked IPs, so we can get serious performance problems in such progress.

Also, I think we do not need to keep all such parasite networks in our DB, but instead a priory reject all requests from them. From my experience there are mobile/customer operators from some countries involved in SPAM and DDoS attacks.
Currently I'm working in my spare time on new module which should allow to reject requests to your site based on whole country (countries) IP subnets. Subnets per country are kept in simple text files.

Admin page draft screenshot for reference:

![ip_deny_country](https://github.com/backdrop-contrib/ip_blocking/assets/11479969/aa8fd4a4-4801-4086-a663-433dc3731e05)

3. Not so clear to me for what you need to edit IP instead of unblocking it and blocking another. 
I prefer to keep things as simple as possible.

Thanks!

**Comment by alanmels** [comment]
Thanks for the prompt reply. I agree with all your arguments. Very promising screenshot. I will wait for your new module to be rolled out. Meanwhile let's close this one.
